File: lib/storage.ts
import { STORAGE_BUCKET } from "@/lib/constants";
import { getSupabaseAdmin } from "@/lib/supabase";

export async function uploadProjectFile(path: string, buffer: Buffer, contentType: string) {
  const supabase = getSupabaseAdmin();
  const { error } = await supabase.storage.from(STORAGE_BUCKET).upload(path, buffer, {
    contentType,
    upsert: true
  });
  if (error) throw error;

  const { data } = supabase.storage.from(STORAGE_BUCKET).getPublicUrl(path);
  return data.publicUrl;
}

export async function fetchStorageBase64(path: string) {
  const supabase = getSupabaseAdmin();
  const { data, error } = await supabase.storage.from(STORAGE_BUCKET).download(path);
  if (error) throw error;
  const arrayBuffer = await data.arrayBuffer();
  return Buffer.from(arrayBuffer).toString("base64");
}

export async function uploadRemoteImage(path: string, url: string) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Remote image download failed: ${response.status}`);
  }
  const buffer = Buffer.from(await response.arrayBuffer());
  return uploadProjectFile(path, buffer, response.headers.get("content-type") ?? "image/png");
}

File: backend/app/storage.py
import base64
import httpx

from .config import get_settings
from .supabase_client import get_supabase


def upload_project_file(path: str, data: bytes, content_type: str) -> str:
    settings = get_settings()
    supabase = get_supabase()
    bucket = settings.supabase_storage_bucket
    storage = supabase.storage.from_(bucket)
    storage.upload(
        path=path,
        file=data,
        file_options={"content-type": content_type, "upsert": "true"},
    )
    return storage.get_public_url(path)


def fetch_storage_base64(path: str) -> str:
    supabase = get_supabase()
    bucket = get_settings().supabase_storage_bucket
    blob = supabase.storage.from_(bucket).download(path)
    return base64.b64encode(blob).decode("utf-8")


def fetch_storage_bytes(path: str) -> bytes:
    supabase = get_supabase()
    bucket = get_settings().supabase_storage_bucket
    return supabase.storage.from_(bucket).download(path)


async def upload_remote_image(path: str, url: str) -> str:
    async with httpx.AsyncClient(timeout=60.0) as client:
        resp = await client.get(url)
        resp.raise_for_status()
    content_type = resp.headers.get("content-type", "image/png")
    return upload_project_file(path, resp.content, content_type)
